### PR TITLE
[TOPIC-BLE-LLCP] Initial demo implementation of LLCP system framework

### PR DESCRIPTION
**Pull request by thoh-ot**
I havn't fixed the line width problems

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+
+	tx = (struct node_tx *) mem_acquire(&mem_tx.free);
+	return tx;
+}
+
+/*
+ * Version Exchange Procedure Helper
+ */
+
+static void pdu_assemble_version_ind(struct pdu_data *pdu)
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c]: pdu_version_ind_encode?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+		pdu_assemble_version_ind(pdu);
+		ctx->opcode = PDU_DATA_LLCTRL_TYPE_VERSION_IND;
+		break;
+	default:
+		/* Unknown procedure */
+		LL_ASSERT(0);
+	}
+
+	/* Enqueue LL Control PDU towards LLL */
+	ull_tx_q_enqueue_ctrl(conn->tx_q, tx);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c]: rename to ull_tx_q_ctrl_enqueue

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+}
+
+
+static void lp_comm_st_idle(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt)
+{
+	switch (evt) {
+	case LP_COMMON_EVT_PREPARE:
+		if (!tx_alloc_peek() || ctx->pause || ctx->collision) {
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c]: Why ctx->collision leads to WAIT_TX?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+		/* Unknown procedure */
+		LL_ASSERT(0);
+	}
+
+	lp_comm_execute_fsm(conn, ctx, LP_COMMON_EVT_PREPARE);
+}
+
+static void lr_act_complete(struct ull_cp_conn *conn)
+{
+	/* Dequeue pending connuest that just completed */
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c]: connuest ?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+{
+	/**/
+	mem_init(mem_ctx.pool, sz_mem_ctx, ncount_mem_ctx, &mem_ctx.free);
+	mem_init(mem_tx.pool, sz_mem_tx, ncount_mem_tx, &mem_tx.free);
+}
+
+void ull_cp_conn_init(struct ull_cp_conn *conn)
+{
+	/**/
+	conn->local.state = LR_STATE_DISCONNECT;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c]: LR_STATE_DISCONNECT? Shouldn't it be something LR_STATE_IDLE?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	/* Tx Queue Reference
+	 * TODO(thoh):
+	 *	Temporary solution to contain the changes.
+	 *	To be transitioned to ll_conn. */
+	struct ull_tx_q *tx_q;
+
+	/* Local Request */
+	struct {
+		sys_slist_t pend_proc_list;
+		u8_t state;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.h]: What values does it hold?

File: tests/bluetooth/ctrl_llcp/src/main.c [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+void test_init(void)
+{
+	ull_tx_q_init(&tx_q);
+
+	ull_cp_conn_init(&conn);
+	zassert_equal(conn.local.state, LR_STATE_DISCONNECT, NULL);
+
+	conn.tx_q = &tx_q;
+
+	ull_cp_init();
```
Comment (cvinayak) [inline tests/bluetooth/ctrl_llcp/src/main.c]: Move this before `ull_cp_conn_init` ?

**Review by cvinayak — CHANGES_REQUESTED** [review]
Some comments. I would like to spend another day at this review.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+#include <soc.h>
+#include "hal/debug.h"
+
+/* LLCP Local Procedure Common FSM states */
+enum {
+	LP_COMMON_STATE_IDLE,
+	LP_COMMON_STATE_WAIT_TX,
+	LP_COMMON_STATE_WAIT_RX,
+	LP_COMMON_STATE_WAIT_NTF,
+};
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: These are the logical states of the Local Procedure Common FSM. You can only be in one of the states at a time.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+
+/* LLCP Local Request FSM Event */
+enum {
+	LR_EVT_PREPARE,
+	LR_EVT_COMPLETE,
+	LR_EVT_CONNECT,
+	LR_EVT_DISCONNECT,
+};
+
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: `LR_EVT_` belong to the Local Request FSM and `LP_COMMON_EVT_` belong to Local Procedure Common FSM. `LP_COMMON_EVT_PREPARE` is to be used while the Local Procedure Common FSM is in the `LP_COMMON_STATE_WAIT_TX` to get a Tx buffer.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+/* LLCP Procedure Context */
+struct proc_ctx {
+	/* Must be the first for sys_slist to work */
+	sys_snode_t node;
+
+	/**/
+	u8_t proc;
+	u8_t state;
+
+	u8_t opcode;
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: Yes, it's a premature optimization `opcode = f(proc, state)`. By storing the expected opcode you avoid having to maintain a lookup mechanism based on `proc` and `state`.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+
+	int collision;
+	int pause;
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: This is not an optimized implementation, but yes they should probably be bitfields in the long run.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+};
+
+/* LLCP Memory Pool Descriptor */
+struct mem_pool {
+	void *free;
+	u8_t *pool;
+};
+
+#define PROC_CTX_BUF_SIZE       WB_UP(sizeof(struct proc_ctx))
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: It's from sys/util.h to round up on a word boundary.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+static struct mem_pool mem_ctx = { .pool = buffer_mem_ctx };
+static const size_t sz_mem_ctx = PROC_CTX_BUF_SIZE;
+static const u32_t ncount_mem_ctx = 1;
+
+#else
+
+/* LLCP Allocations (unittest) */
+
+extern const size_t sz_mem_tx;
+extern const u32_t ncount_mem_tx;
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: No, the idea is that the unit test can specify the size of buffers, but if you make the extern declaration have a size then the definition in the unit test have a different size, the compiler will complain. That is also why the unit test part does not have any buffer definitions as what matters is the information stored in `mem_tx` and `mem_ctx`.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 1 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: Fine with me.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+ */
+
+static void lp_comm_tx(struct ull_cp_conn *conn, struct proc_ctx *ctx)
+{
+	struct node_tx *tx;
+	struct pdu_data *pdu;
+
+	/* Allocate tx node */
+	tx = tx_alloc();
+	LL_ASSERT(tx);
```
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: No, there is no semaphores involved. I have not shown how the wait for free tx buffer works in this PR. `lp_comm_tx()` is only called if `tx_alloc_peek()` is true in `lp_comm_st_idle()`.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 2 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: For me it makes more sense as it is. It's name divides into three parts: the namespace `ull_tx_q`, the action `enqueue` and the type of element `ctrl`.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 4 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: This should be "request"

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 5 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: `LR_STATE_DISCONNECT` is a state yo make sure that you cannot perform any more interaction before you bring it back into `LR_STATE_IDLE`.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h (hunk 6 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.h]: Values from `LR_STATE_`

File: tests/bluetooth/ctrl_llcp/src/main.c (hunk 7 above)
Comment (thoh-ot) [reply tests/bluetooth/ctrl_llcp/src/main.c]: Yes, that would probably make more sense.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h (hunk 6 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.h]: ok

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: Is there a possibility that in LP_COMMON_STATE_WAIT_NTF and a new PDU arrives (example, multiple CONN_UPDATE_IND), how is this to be handled?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: ok

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: fine with me.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: ok

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: in the unit test you include the c file, you can re-define the size macros as required before the inclusion.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: ok

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: ??

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 5 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: ok

**Review by cvinayak — DISMISSED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: That would be a possibility, let me look in to it.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 3 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: This is for the case where prior to the local procedure sending the first pdu, you already know that there is a remote procedure running that would give a collision, so WAIT_TX is wait state until the colliding procedure is finished.

**Review by kruithofa — APPROVED** [review]

**Comment by joerchan** [comment]
@cvinayak  Merged with DNM label?

**Pull request merged**